modular/gater/admin_handler.go

```diff
@@ -413,3 +418,207 @@ func (g *GateModular) replicateHandler(w http.ResponseWriter, r *http.Request) {
 	}
 	log.CtxDebugw(reqCtx.Context(), "succeed to replicate piece")
 }
+
+// recoveryPrimaryHandler handles the query for recovery request from secondary SP or primary SP.
+// if it is used to recovery primary SP and the handler SP is the corresponding secondary,
+// it returns the EC piece data stored in the secondary SP for the requested object.
+// if it is used to recovery secondary SP and the handler is the corresponding primary SP,
+// it directly returns the EC piece data of the secondary SP.
+func (g *GateModular) recoveryPrimaryHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		err         error
+		reqCtx      *RequestContext
+		recoveryMsg []byte
+	)
+	defer func() {
+		reqCtx.Cancel()
+		if err != nil {
+			reqCtx.SetError(gfsperrors.MakeGfSpError(err))
+			reqCtx.SetHttpCode(int(gfsperrors.MakeGfSpError(err).GetHttpStatusCode()))
+			MakeErrorResponse(w, gfsperrors.MakeGfSpError(err))
+		} else {
+			reqCtx.SetHttpCode(http.StatusOK)
+		}
+		log.CtxDebugw(reqCtx.Context(), reqCtx.String())
+	}()
+
+	log.CtxDebugw(reqCtx.Context(), "recoveryPrimaryHandler begin1")
+	// ignore the error, because the recovery request only between SPs, the request
+	// verification is by signature of the RecoveryTask
+	reqCtx, _ = NewRequestContext(r, g)
+
+	recoveryMsg, err = hex.DecodeString(r.Header.Get(GnfdRecoveryMsgHeader))
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to parse recovery header",
+			"header", r.Header.Get(GnfdRecoveryMsgHeader))
+		err = ErrDecodeMsg
+		return
+	}
+
+	recoveryTask := gfsptask.GfSpRecoveryPieceTask{}
+	err = json.Unmarshal(recoveryMsg, &recoveryTask)
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to unmarshal receive header",
+			"receive", r.Header.Get(GnfdReceiveMsgHeader))
+		err = ErrDecodeMsg
+		return
+	}
+
+	// check signature consistent
+	taskSignature := recoveryTask.GetSignature()
+	signatureAddr, pk, err := RecoverAddr(recoveryTask.GetSignBytes(), taskSignature)
+	if err != nil {
+		log.CtxErrorw(reqCtx.Context(), "failed to  get recover task address", "error:", err)
+		err = ErrSignature
+		return
+	}
+
+	log.CtxDebugw(reqCtx.Context(), "get recovery piece request from sp:", signatureAddr)
+	if !secp256k1.VerifySignature(pk.Bytes(), recoveryTask.GetSignBytes(), taskSignature[:len(taskSignature)-1]) {
+		log.CtxErrorw(reqCtx.Context(), "failed to verify recovery task signature")
+		err = ErrSignature
+		return
+	}
+
+	// handle request from primary SP to recovery segment {
+	objectInfo := recoveryTask.GetObjectInfo()
+	if objectInfo == nil {
+		log.CtxErrorw(reqCtx.Context(), "recovery task params error")
+		err = ErrInvalidHeader
+		return
+	}
+
+	if objectInfo.RedundancyType != storagetypes.REDUNDANCY_EC_TYPE {
+		log.CtxErrorw(reqCtx.Context(), "recovery redundancy type only support EC")
+		err = ErrRecoveryRedundancyType
+		return
+	}
+
+	chainObjectInfo, bucketInfo, params, err := getObjectChainMeta(reqCtx, g.baseApp, objectInfo.ObjectName, objectInfo.BucketName)
```

Comment:
g.baseapp has func get object info and object info

Reply:
the getObjectMeta fucntion called g.baseapp.Consensus().QueryObject. and  g.baseapp.Consensus().QueryBucket inside it